build_tools/cmake/iree_plugin_register.cmake

```diff
@@ -0,0 +1,133 @@
+# it will be runtime/plugins/hal/drivers/${NAME}
+#
+# Args:
+#  NAME: The name of the hal driver.
+#  OPTIONAL: Whether the driver is optional and it is not an error if it is
+#    disabled.
+#  SOURCE_DIR: Source directory to include if the plugin is enabled.
+#  BINARY_DIR: Binary directory corresponding to the SOURCE_DIR.
+#  DRIVER_TARGET: CMake target to be linked to.
+#  REGISTER_FN: C symbol of the registration function.
+function(iree_register_external_hal_driver)
+  cmake_parse_arguments(
+    _RULE
+    "OPTIONAL"
+    "NAME;SOURCE_DIR;BINARY_DIR;DRIVER_TARGET;REGISTER_FN"
+    ""
+    ${ARGN}
+  )
+
+  # Normalize binary dir.
+  if(NOT _RULE_BINARY_DIR)
+    set(_RULE_BINARY_DIR "${_RULE_NAME}")
+  endif()
+  cmake_path(ABSOLUTE_PATH _RULE_BINARY_DIR 
+    BASE_DIRECTORY "${IREE_BINARY_DIR}/runtime/plugins/hal/drivers")
+
+  string(TOUPPER "${_RULE_NAME}" _NAME_SPEC)
+  string(REGEX REPLACE "-" "_" _NAME_SPEC ${_NAME_SPEC})
+
+  get_property(_prev_available GLOBAL PROPERTY IREE_EXTERNAL_HAL_DRIVERS_AVAILABLE)
+  if(_RULE_NAME IN_LIST _prev_available)
+    message(SEND_ERROR "iree_register_external_hal_driver(NAME ${_RULE_NAME}) called more than once")
+  endif()
+
+  set_property(GLOBAL APPEND PROPERTY IREE_EXTERNAL_HAL_DRIVERS_AVAILABLE "${_RULE_NAME}")
+  if(_RULE_OPTIONAL)
+    set_proeprty(GLOBAL_PROPERTY "IREE_EXTERNAL_${_NAME_SPEC}_HAL_DRIVER_OPTIONAL" TRUE)
+  endif()
+  set_property(GLOBAL PROPERTY "IREE_EXTERNAL_${_NAME_SPEC}_HAL_DRIVER_SOURCE_DIR"
+    "${_RULE_SOURCE_DIR}")
+  set_property(GLOBAL PROPERTY "IREE_EXTERNAL_${_NAME_SPEC}_HAL_DRIVER_BINARY_DIR"
+    "${_RULE_BINARY_DIR}")
+  set_property(GLOBAL PROPERTY "IREE_EXTERNAL_${_NAME_SPEC}_HAL_DRIVER_TARGET"
+    "${_RULE_DRIVER_TARGET}")
+  set_property(GLOBAL PROPERTY "IREE_EXTERNAL_${_NAME_SPEC}_HAL_DRIVER_REGISTER"
+    "${_RULE_REGISTER_FN}")
+endfunction()
+
+# iree_include_cmake_plugin_dirs()
+# Configures all in-tree and out-of-tree plugins based on global settings.
+function(iree_include_cmake_plugin_dirs)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "LOG_LABEL;BINARY_DIR;PLUGIN_CMAKE_FILE"
+    ""
+    ${ARGN}
+  )
+
+  file(MAKE_DIRECTORY "${_RULE_BINARY_DIR}")
+
+  # Generate an include() in BINARY_DIR/CMakeLists.txt to sub-include each
+  # plugin cmake file.
+  set(_contents)
+  foreach(_d ${IREE_CMAKE_BUILTIN_PLUGIN_PATHS} ${IREE_CMAKE_PLUGIN_PATHS})
+    cmake_path(ABSOLUTE_PATH _d BASE_DIRECTORY "${IREE_SOURCE_DIR}" NORMALIZE)
+    set(_f "${_d}/${_RULE_PLUGIN_CMAKE_FILE}")
+    if(NOT EXISTS "${_f}")
+      continue()
+    endif()
+    string(APPEND _contents "include(${_f})\n")
+  endforeach()
+  file(CONFIGURE OUTPUT "${_RULE_BINARY_DIR}/CMakeLists.txt"
+    CONTENT "${_contents}" @ONLY
+  )
+  unset(_contents)
+  unset(_d)
+  unset(_f)
+
+  message(STATUS "Configuring IREE ${_RULE_LOG_LABEL} plugins")
+  add_subdirectory("${_RULE_BINARY_DIR}" "${_RULE_BINARY_DIR}")
```

Comment:
We could add a message _after_ configuring too, to make it clear when the nesting finishes

![image](https://github.com/openxla/iree/assets/4010439/fb2a1a99-33ad-416b-9d5a-0c7dc32dd64c)

Maybe there's a way to have CMake indent messages...?

```
// like this
-- Configuring IREE compiler plugins
--   Registering static linked compiler plugin 'torch_iree' (torch-iree::registration)
   ^^ indent
```

Reply:
I don't think those are all coming from the same place, they just happen to be encountered sequentially while loading CMake files...

Reply:
Ah, there is something: https://cmake.org/cmake/help/latest/variable/CMAKE_MESSAGE_INDENT.html 

I've seen a few projects use mechanisms like this for some ergonomics.